Insert Ignore Option

Zugwalt opened the pull request:
Sometimes when inserting into a MySQL table you want to ignore duplicate keys. This PR adds that option to `create()` and `createBulk()`.

Rather then doing something like `INSERT... SELECT` with a `LEFT JOIN` back to the table and doing `WHERE id IS NULL`, we can leverage MySQL's `ON DUPLICATE KEY`. This approach was done as `INSERT IGNORE` ignores _all_ errors, not just duplicate keys.

jonathan-fulton commented on lib/mysqlDao.js:
```diff
@@ -42,11 +42,12 @@ class MysqlDao {
      */
-    create(createArgs) {
-        const sql = this._queryHelper.create(createArgs);
+    create(createArgs, ignoreOnDuplicateKey) {
```
I would probably create an "options" object argument with ignoreOnDuplicateKey being one of the possible options.  Just in case we need to add more stuff in the future.

Zugwalt replied:
Ok will do!

jonathan-fulton commented on lib/mysqlDao.js:
```diff
@@ -65,10 +66,11 @@ class MysqlDao {
      */
-    createBulk(createArgsArr) {
-        const sql = this._queryHelper.createBulk(createArgsArr);
+    createBulk(createArgsArr, ignoreOnDuplicateKey) {
```
Same here.  Options arg.

jonathan-fulton commented on lib/mysqlDaoQueryHelper.js:
```diff
@@ -31,32 +31,35 @@ class MysqlDaoQueryHelper {
-                .setFields(cleanValues, {dontQuote: true})
-                .toString();
-        return sql;
+    create(createArgs, ignoreOnDuplicateKey) {
```
Pass along options arg.

jonathan-fulton commented on lib/mysqlDaoQueryHelper.js:
```diff
@@ -31,32 +31,35 @@ class MysqlDaoQueryHelper {
+     * @param {boolean} [ignoreOnDuplicateKey] if true, insert will ignore duplicate keys
      */
-    createBulk(createArgsArray) {
+    createBulk(createArgsArray, ignoreOnDuplicateKey) {
```
And here.

jonathan-fulton commented on lib/mysqlDaoQueryHelper.js:
```diff
@@ -31,32 +31,35 @@ class MysqlDaoQueryHelper {
+            //INSERT ... IGNORE actually ignores ALL errors, not just duplicate key, so best practice is to "update"
+            // a duplicate setting a column to itself
+            const firstField = Object.keys(cleanValues[0])[0];
+            sqlBuilder.onDupUpdate(firstField, firstField, {dontQuote: true})
```
Not 100% sure, but we may need to do something similar to what we did for upsertBulk.  We had to do 

``` 
INSERT INTO (col1, col2,...)
VALUES 
  ('val1_1', 'val2_1',...),
  ('val1_2, 'val2_2',...)
ON DUPLICATE KEY UPDATE
  col1 = VALUES(col1),
  col2 = VALUES(col2)
```

The natural col1 = col1 ends up using the value from the first set of insert values or something like that.  Might be worth pinging Jason about it b/c he did upsertBulk.  It looks like this might not be an issue based on the unit test but worth confirming.

Zugwalt replied:
The SQL it generates is fine. `col1 = col1` basically well set it to the row value, and MySQL is smart enough to know it doesn't even need to touch the row.

Zugwalt replied:
Negative, MySQL is smart enough to know no update actually happens. (confirmed by doing a test myself)

cdeveas commented on lib/mysqlDaoQueryHelper.js:
```diff
@@ -31,32 +31,35 @@ class MysqlDaoQueryHelper {
+        if (ignoreOnDuplicateKey) {
+            //INSERT ... IGNORE actually ignores ALL errors, not just duplicate key, so best practice is to "update"
+            // a duplicate setting a column to itself
+            const firstField = Object.keys(cleanValues[0])[0];
```
Is a null reference possible?

Zugwalt replied:
Should not happen with valid use. Would happen only if either the array to insert is empty, or the object is empty. These are technically invalid inputs, and I think both of these will cause errors later down the line anyways.

Zugwalt commented:
> We probably want to switch to an object "options" argument.

Agree! Changed to `opts` argument for future expandability.

> We should add unit tests to test-mysqlDaoQueryHelper.

Agree! Added.

> Confirm that we don't need col1 = VALUES(col1) in the ON DUP KEY UPDATE clause.

Not needed, we simply need `col1 = col1` to effectively ignore the insert in a more robust way than `INSERT IGNORE`. (confirmed with unit tests; also for discussion see: http://stackoverflow.com/questions/2366813/on-duplicate-key-ignore)

jonathan-fulton commented:
Merged.
